# task 6.1sourcing open data

#    project name : New York bike riders Rental analysis

# 01. Importing  libraries

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
import calendar


#  02. Importing citybike dataset
     
     step 2.1 creating a folderpath for the dataset 
     
     step 2.2  Importing the citybike dataset for exploratory analysis

In [2]:
# creating a folder path for city bike dataset ;
path =  r"C:\Users\solza\Downloads\02-11-2023"

In [3]:
# import citybike dataset
df = pd.read_csv(r"C:\Users\solza\Downloads\02-11-2023\citybike.csv")

# data profiling and Integrity
  
 # step 1 data profiling 
 
     1.1 dataframe dimension and shape
     1.2 summary statistics prior to data set clean up
     1,3 cheking dataframe column details
     1.4 checking dataframe variables
 
 # step 2 data integrity
 
    2.1 data frame uniquness : finding duplicates
    2.2 checking for dataframe mixed data frame
    2.3 data frame completness : finding missing values
    2.4 data set timelines : collected septemper 2013
    
# step 3 changes performed

    3.1 change 'gender' column as 'rider gender'
    3.2 delete unnecessary column, 'birth year', 'trip_id' and 'gender'
    3.3 run descriptive statistics after clean up
    
    

# step 1 data profiling 

In [4]:
# checking for data frame shape;

df.shape
 

(50000, 18)

 ##  Note : original data has 50,000 records (rows) and 18 (columns) in total

In [5]:
# checking data frame dimentions;
df.head()

trip_id  bike_id weekday  start_hour       start_time  start_station_id  \
0  LnQzQk    16013     Mon          18   9/9/2013 18:18               523   
1  IL9boN    15230     Thu          18  9/12/2013 18:38               257   
2  46clGB    17942     Wed          19  9/18/2013 19:44               479   
3  v7vdFt    19683     Sat          11  9/28/2013 11:54               527   
4  VGBsb5    18024     Sat          18   9/7/2013 18:08               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

          end_time  end_station_id     end_station_name  end_station_latitude  \
0   9/9/2013 18:35             334      W 20 St & 7 Ave             40.742388   
1  9/12/2013 18:48             236  St Marks Pl & 2 Ave             40.728419   
2  9/18/2013 19:50             513     W 56 St & 10 Ave             40.768254   
3  9/28/2013 12:03             441      E 52 St & 2 Ave             40.756014   
4   9/7/2013 18:46             476      E 31 St & 3 Ave             40.743943   

   end_station_longitude  trip_duration      subscriber  birth_year  gender  
0             -73.997262            993      Subscriber      1968.0       2  
1             -73.987140            581      Subscriber      1983.0       1  
2             -73.988639            361      Subscriber      1989.0       1  
3             -73.967416            561      Subscriber      1988.0       2  
4             -73.979661           2296  Non-Subscriber         NaN       0

###  note: column 4 trip duration is defined as 2296 while start time is 18:08 and end time is 18:46. this indicates 36  minutes. 2296 is not identified as second or microseconds.

In [6]:
# step 1.2: running descriptive stastics prior to clean up;
df.describe()

bike_id    start_hour  start_station_id  start_station_latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       start_station_longitude  end_station_id  end_station_latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       end_station_longitude  trip_duration    birth_year        gender  
count           50000.000000   50000.000000  43021.000000  50000.000000  
mean              -73.991351     838.982900   1975.627786      1.073540  
std                 0.012569     573.663997     11.089001      0.589389  
min               -74.017134      60.000000   1899.000000      0.000000  
25%               -74.001547     417.000000   1968.000000      1.000000  
50%               -73.991218     672.000000   1978.000000      1.000000  
75%               -73.982050    1112.000000   1984.000000      1.000000  
max               -73.950048    2697.000000   1997.000000      2.000000

### Note : seems 'trip_duration' is in second or microseconds

In [7]:
#Copying Descriptive Statistics to clipboard before Cleanup;
df.describe().to_clipboard()

In [8]:
# checking data frame columns detail:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   trip_id                  50000 non-null  object 
 1   bike_id                  50000 non-null  int64  
 2   weekday                  50000 non-null  object 
 3   start_hour               50000 non-null  int64  
 4   start_time               50000 non-null  object 
 5   start_station_id         50000 non-null  int64  
 6   start_station_name       50000 non-null  object 
 7   start_station_latitude   50000 non-null  float64
 8   start_station_longitude  50000 non-null  float64
 9   end_time                 50000 non-null  object 
 10  end_station_id           50000 non-null  int64  
 11  end_station_name         50000 non-null  object 
 12  end_station_latitude     50000 non-null  float64
 13  end_station_longitude    50000 non-null  float64
 14  trip_duration         

In [9]:
# checking for dataframe variables and data types
df.dtypes

trip_id                     object
bike_id                      int64
weekday                     object
start_hour                   int64
start_time                  object
start_station_id             int64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_time                    object
end_station_id               int64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
trip_duration                int64
subscriber                  object
birth_year                 float64
gender                       int64
dtype: object

In [10]:
# copying data type to clipboard
df.dtypes.to_clipboard()

## step 2 data integrity 

In [11]:
# finding duplicates:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
49995    False
49996    False
49997    False
49998    False
49999    False
Length: 50000, dtype: bool

### findings : no duplicates exist 

In [12]:
# Checking for Mixed-Data Types; 

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

###  findings : no mixed data type 

In [13]:
# checking for missing values
df.isnull().sum()

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
dtype: int64

###  Findings : ' birth_year' column has 6979 ( 14%) missing values

### Notes :  the columns with 'birth_year will be droped since it is okey to proceed the analysis with out birth of the riders.


In [14]:
#create new data frame for the actuall missing value under 'birth_year';

df_nan = df[df['birth_year'].isnull() == True]

In [15]:
#viewing the actual missing values under 'birth_year';
df_nan

trip_id  bike_id weekday  start_hour       start_time  start_station_id  \
4      VGBsb5    18024     Sat          18   9/7/2013 18:08               521   
10     aZsCdK    15732     Sat          20   9/7/2013 20:01               457   
17     VqhF3S    15651     Mon          18  9/30/2013 18:32               363   
26     9yZrq6    17506     Sun          16   9/1/2013 16:18               389   
46     t1KBiF    19206     Sun          22   9/1/2013 22:49               426   
...       ...      ...     ...         ...              ...               ...   
49986  8wDcq8    19293     Sun          22  9/29/2013 22:00               469   
49988  C0ivmQ    18118     Sun          12  9/22/2013 12:19               352   
49993  cvgHF4    18458     Wed          19  9/25/2013 19:01               284   
49995  Fb2bmC    17988     Sat          18  9/21/2013 18:39               479   
49998  xxacSt    16763     Sat          13   9/7/2013 13:25               271   

           start_station_name  start_station_latitude  \
4             8 Ave & W 31 St               40.750450   
10         Broadway & W 58 St               40.766953   
17             West Thames St               40.708347   
26        Broadway & Berry St               40.710446   
46      West St & Chambers St               40.717548   
...                       ...                     ...   
49986      Broadway & W 53 St               40.763441   
49988         W 56 St & 6 Ave               40.763406   
49993   Greenwich Ave & 8 Ave               40.739017   
49995         9 Ave & W 45 St               40.760193   
49998  Ashland Pl & Hanson Pl               40.685282   

       start_station_longitude         end_time  end_station_id  \
4                   -73.994811   9/7/2013 18:46             476   
10                  -73.981693   9/7/2013 20:23             512   
17                  -74.017134  9/30/2013 19:01             514   
26                  -73.965251   9/1/2013 16:26             389   
46                  -74.013221   9/1/2013 23:06             247   
...                        ...              ...             ...   
49986               -73.982681  9/29/2013 22:17             444   
49988               -73.977225  9/22/2013 12:29             464   
49993               -74.002638  9/25/2013 19:19             329   
49995               -73.991255  9/21/2013 20:08             478   
49998               -73.978058   9/7/2013 14:14             373   

                   end_station_name  end_station_latitude  \
4                   E 31 St & 3 Ave             40.743943   
10                  W 29 St & 9 Ave             40.750073   
17                 12 Ave & W 40 St             40.760875   
26              Broadway & Berry St             40.710446   
46           Perry St & Bleecker St             40.735354   
...                             ...                   ...   
49986            Broadway & W 24 St             40.742354   
49988               E 56 St & 3 Ave             40.759345   
49993     Greenwich St & N Moore St             40.720434   
49995              11 Ave & W 41 St             40.760301   
49998  Willoughby Ave & Walworth St             40.693317   

       end_station_longitude  trip_duration      subscriber  birth_year  \
4                 -73.979661           2296  Non-Subscriber         NaN   
10                -73.998393           1313  Non-Subscriber         NaN   
17                -74.002777           1751  Non-Subscriber         NaN   
26                -73.965251            478  Non-Subscriber         NaN   
46                -74.004831           1016  Non-Subscriber         NaN   
...                      ...            ...             ...         ...   
49986             -73.989151           1029  Non-Subscriber         NaN   
49988             -73.967597            611  Non-Subscriber         NaN   
49993             -74.010206           1084  Non-Subscriber         NaN   
49995             -73.998842           2697  Non-Subscriber 

##  step 3.change performed 

In [16]:
#defining 'gender' type characterstics;
Rider_Gender = []

for value in df["gender"]:
    if value == 1:
        Rider_Gender.append("Male")
    elif value == 2:
        Rider_Gender.append("Female")
    else:
        Rider_Gender.append("Unknown")
  

In [17]:
df['Rider_Gender'] = Rider_Gender

In [18]:
df['Rider_Gender'].value_counts(dropna = False)

Male       32361
Female     10658
Unknown     6981
Name: Rider_Gender, dtype: int64

In [19]:
# cheking for new column;
df.head()

trip_id  bike_id weekday  start_hour       start_time  start_station_id  \
0  LnQzQk    16013     Mon          18   9/9/2013 18:18               523   
1  IL9boN    15230     Thu          18  9/12/2013 18:38               257   
2  46clGB    17942     Wed          19  9/18/2013 19:44               479   
3  v7vdFt    19683     Sat          11  9/28/2013 11:54               527   
4  VGBsb5    18024     Sat          18   9/7/2013 18:08               521   

        start_station_name  start_station_latitude  start_station_longitude  \
0          W 38 St & 8 Ave               40.754666               -73.991382   
1  Lispenard St & Broadway               40.719392               -74.002472   
2          9 Ave & W 45 St               40.760193               -73.991255   
3          E 33 St & 1 Ave               40.743156               -73.974347   
4          8 Ave & W 31 St               40.750450               -73.994811   

          end_time  end_station_id     end_station_name  end_station_latitude  \
0   9/9/2013 18:35             334      W 20 St & 7 Ave             40.742388   
1  9/12/2013 18:48             236  St Marks Pl & 2 Ave             40.728419   
2  9/18/2013 19:50             513     W 56 St & 10 Ave             40.768254   
3  9/28/2013 12:03             441      E 52 St & 2 Ave             40.756014   
4   9/7/2013 18:46             476      E 31 St & 3 Ave             40.743943   

   end_station_longitude  trip_duration      subscriber  birth_year  gender  \
0             -73.997262            993      Subscriber      1968.0       2   
1             -73.987140            581      Subscriber      1983.0       1   
2             -73.988639            361      Subscriber      1989.0       1   
3             -73.967416            561      Subscriber      1988.0       2   
4             -73.979661           2296  Non-Subscriber         NaN       0   

  Rider_Gender  
0       Female  
1         Male  
2         Male  
3       Female  
4      Unknown

In [20]:
# deleting unnecessary column;
# 'trip_ID' and 'birth_year' will not be needed for this analysis
df = df.drop(columns = ['trip_id', 'birth_year',])

In [21]:
df

bike_id weekday  start_hour       start_time  start_station_id  \
0        16013     Mon          18   9/9/2013 18:18               523   
1        15230     Thu          18  9/12/2013 18:38               257   
2        17942     Wed          19  9/18/2013 19:44               479   
3        19683     Sat          11  9/28/2013 11:54               527   
4        18024     Sat          18   9/7/2013 18:08               521   
...        ...     ...         ...              ...               ...   
49995    17988     Sat          18  9/21/2013 18:39               479   
49996    18458     Sat          17  9/14/2013 17:32               261   
49997    15054     Sat          18   9/7/2013 18:08               402   
49998    16763     Sat          13   9/7/2013 13:25               271   
49999    16794     Mon          21  9/23/2013 21:54               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude         end_time  end_station_id  \
0                   -73.991382   9/9/2013 18:35             334   
1                   -74.002472  9/12/2013 18:48             236   
2                   -73.991255  9/18/2013 19:50             513   
3                   -73.974347  9/28/2013 12:03             441   
4                   -73.994811   9/7/2013 18:46             476   
...                        ...              ...             ...   
49995               -73.991255  9/21/2013 20:08             478   
49996               -73.983625  9/14/2013 17:52             151   
49997               -73.989551   9/7/2013 18:25             524   
49998               -73.978058   9/7/2013 14:14             373   
49999               -73.994618  9/23/2013 22:00             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  gender  \
0                 -73.997262            993      Subscriber       2   
1                 -73.987140            581      Subscriber       1   
2                 -73.988639            361      Subscriber       1   
3                 -73.967416            561      Subscriber       2   
4                 -73.979661           2296  Non-Subscriber       0   
...                      ...            ...             ...     ...   
49995             -73.998842           2697  Non-Subscriber       0   
49996             -73.997203           1191      Subscriber       2   
49997             -73.983169           1050      Subscriber       2   
49998             -73.953820           2697  Non-Subscriber       0   
49999             -73.993722            370      Subscriber       2   

      Rider_Gender  
0           Female  


In [22]:
#Copying Descriptive Statistics to clipboard After Cleanup;

df.describe().to_clipboard()

## 4 export cleaned 'citybike' dataset

In [23]:
#Exporting Cleaned 'df_Citibike_Clean' Dataset;
df.to_csv(os.path.join(path, 'Prepared', 'citybike_cleaned.csv'))